File: DFS/282.Expression_Add_Operators/282.Expression_Add_Operators.cpp
class Solution {
public:
    vector<string> addOperators(string num, int target) {
        DFS(num, target, "", 0, 0, 0);
        
        return res;
    }
    
    void DFS(string &num, int &target, string cur_str, long pre_num, long cur_val, int pos) {
        if (pos == num.size()) {
            if (cur_val == target) {
                res.push_back(cur_str);
            }
            return;
        }
        
        for (int i = pos + 1; i <= num.size(); i++) {
            if (num[pos] == '0' && i > pos + 1) return;
            
            string add_str = num.substr(pos, i - pos);
            long add_int = stol(add_str);

            if (pos == 0) {
                DFS(num, target, add_str, add_int, add_int, i);
            }
            else {
                DFS(num, target, cur_str + "+" + add_str, add_int, cur_val + add_int, i);
                DFS(num, target, cur_str + "-" + add_str, -add_int, cur_val - add_int, i);
                DFS(num, target, cur_str + "*" + add_str, pre_num * add_int, cur_val - pre_num + pre_num * add_int, i);
                }
        }
    }
private:
    vector<string> res;
};

File: DFS/282.Expression_Add_Operators/282.Expression_Add_Operators.py
class Solution:
    def addOperators(self, num: str, target: int) -> List[str]:
        res = []
        n = len(num)
        
        def DFS(cur_str, pre_num, cur_val, pos) :
            if pos == n :
                if cur_val == target :
                    res.append(cur_str)
                return
            
            for i in range(pos + 1, n + 1) :
                if num[pos] == '0' and i > pos + 1 :
                    return
                    
                add_str = num[pos:i]
                add_num = int(add_str)
                
                if pos == 0 :
                    DFS(add_str, add_num, add_num, i)
                else :
                    DFS(cur_str + "+" + add_str, add_num, cur_val + add_num, i)
                    DFS(cur_str + "-" + add_str, -add_num, cur_val - add_num, i)
                    DFS(cur_str + "*" + add_str, pre_num * add_num, cur_val - pre_num + pre_num * add_num, i)
                
        DFS("", 0, 0, 0)
        
        return res